File: 151.翻转字符串里的单词.py
#
# @lc app=leetcode.cn id=151 lang=python3
#
# [151] 翻转字符串里的单词
#

# @lc code=start
class Solution:
    def reverseWords(self, s: str) -> str:
        k = 0   # k用来记录空格符，因为字符串中可能含有多个连续的空格
        n = len(s)
        for i in range(n):        #从前往后遍历
            while (i<n and s[i] == " "):     #遇到第i个字符是空格，则跳过
                i+=1
            if i == n:
                break              #字符串遍历到头了
            j = i
            while j<n and j!=' ':
                j+=1     #s[i,j)表示一个单词
            s1 = "".join(list(reversed(s[i:j])))
            print(s1)
            if k:
                s[k] = ' '
            while j<i:
                s[k] = s[j]
                j+=1
                k+=1
                       
            
                
                

        
# @lc code=end


'''
思路：
1.先按顺序将每个单词翻转
2，再将整个字符串翻转
'''

File: 字符串/151.翻转字符串里的单词.cpp
/*
 * @lc app=leetcode.cn id=151 lang=cpp
 *
 * [151] 翻转字符串里的单词
 */

// @lc code=start
class Solution {
public:
    string reverseWords(string s) {
        reverse(s.begin(), s.end());                        //整体反转
        int start = 0, end = s.size() - 1;
        while (start < s.size() && s[start] == ' ') start++;//首空格
        while (end >= 0 && s[end] == ' ') end--;            //尾空格
        if (start > end) return "";                         //特殊情况

        for (int r = start; r <= end;) {                    //逐单词反转
            while (s[r] == ' ' && r <= end) r++;            //单词的左空格
            int l = r;
            while (s[l] != ' ' && l <= end) l++;            //单词的右空格
            reverse(s.begin() + r, s.begin() + l);          //将空格间的单词反转
            r = l;
        }
        int tail = start;                                   //处理中间冗余空格
        for (int i = start; i <= end; i++) {
            if (s[i] == ' '&&s[i - 1] == ' ') continue;     //碰到两个连着的空格跳过本次循环
            s[tail++] = s[i];                               //否则改变(复制)字符串
        }
        return s.substr(start, tail - start);               //最终返回 从start开始 tail-start长度的最终字符串
    }
};
// @lc code=end

